### PR TITLE
ZIP-225/244 #3: Create separate bundles for each part of the transaction.

### PR DESCRIPTION
**Pull request by nuttycom**
This builds atop #386 

This  takes inspiration from the Orchard design and creates a separate `Bundle` type for each sub-protocol: transparent, Sprout, Sapling, and the future TZE transaction components.

### CODE REVIEW

File: zcash_primitives/src/transaction/components/sapling.rs [hunk 1]
```diff
@@ -117,7 +211,7 @@ pub struct OutputDescription {
     pub ephemeral_key: jubjub::ExtendedPoint,
     pub enc_ciphertext: [u8; 580],
     pub out_ciphertext: [u8; 80],
-    pub zkproof: [u8; GROTH_PROOF_SIZE],
+    pub zkproof: GrothProofBytes,
```
Comment (str4d) [inline zcash_primitives/src/transaction/components/sapling.rs:214]: This also needs to be typed, since `OutputDescription` can be in an unproven state similar to `SpendDescription`.

File: zcash_primitives/src/transaction/components/sapling.rs [hunk 2]
```diff
@@ -1,36 +1,107 @@
...
+    type Proof = GrothProofBytes;
+    type Signature = ();
+}
+
+#[derive(Debug, Copy, Clone, PartialEq)]
+pub struct Authorized {}
+
+impl Authorization for Authorized {
+    type Proof = GrothProofBytes;
+    type Signature = redjubjub::Signature;
```
Comment (str4d) [inline zcash_primitives/src/transaction/components/sapling.rs:49]: This should be two different types (in preparation for when we move to the `redjubjub` crate), for spendAuth and binding.

File: zcash_primitives/src/transaction/components/sapling.rs [hunk 3]
```diff
@@ -40,49 +111,78 @@ impl std::fmt::Debug for SpendDescription {
...
+    pub fn read_spend_auth_sig<R: Read>(mut reader: R) -> io::Result<Signature> {
+        Signature::read(&mut reader)
+    }
 
-        // Consensus rules (§4.4):
-        // - Canonical encoding is enforced here.
-        // - Signature validity is enforced in SaplingVerificationContext::check_spend()
-        let spend_auth_sig = Some(Signature::read(&mut reader)?);
+    pub fn read<R: Read>(mut reader: R) -> io::Result<Self> {
+        let cv = read_point(&mut reader, "cv")?;
```
Comment (str4d) [inline zcash_primitives/src/transaction/components/sapling.rs:180]: I think I'd prefer to keep the consensus rule comments here, since they apply to the specific variable types, not to the general `read_point` etc. functions (even though it is those functions that enforce the rules enforced "here").

File: zcash_primitives/src/transaction/components/sapling.rs [hunk 4]
```diff
@@ -100,13 +200,7 @@ impl SpendDescription {
...
         self.rk.write(&mut writer)?;
         writer.write_all(&self.zkproof)?;
-        match self.spend_auth_sig {
-            Some(sig) => sig.write(&mut writer),
-            None => Err(io::Error::new(
-                io::ErrorKind::InvalidInput,
-                "Missing spend auth signature",
-            )),
-        }
+        self.spend_auth_sig.write(&mut writer)
```
Comment (str4d) [inline zcash_primitives/src/transaction/components/sapling.rs:203]: 🎉

File: zcash_primitives/src/transaction/components/sapling/builder.rs [hunk 5]
```diff
@@ -113,20 +114,19 @@ impl<P: consensus::Parameters> SaplingOutput<P> {
             to,
             note,
             memo: memo.unwrap_or_else(MemoBytes::empty),
-            _params: PhantomData::default(),
         })
     }
 
-    pub fn build<Pr: TxProver, R: RngCore + CryptoRng>(
+    pub fn build<P: consensus::Parameters, Pr: TxProver, R: RngCore + CryptoRng>(
```
Comment (str4d) [inline zcash_primitives/src/transaction/components/sapling/builder.rs:120]: This would enable using different parameters to construct the builder and then build with it. Undo these changes and re-introduce `P: consensus::Parameters` as a bound on the builder.

Comment (str4d) [reply zcash_primitives/src/transaction/components/sapling/builder.rs:120]: Actually, we could instead make `SaplingOutput` an internal type, and then these changes would be fine. `SaplingOutput` was public because we wanted to enable external transaction builders to add Sapling outputs, but now we have `SaplingBuilder` which provides that functionality with an improved API.

File: zcash_primitives/src/transaction/components/sapling/builder.rs [hunk 6]
```diff
@@ -311,14 +311,7 @@ impl<P: consensus::Parameters> SaplingBuilder<P> {
...
         progress_notifier: Option<&Sender<Progress>>,
-    ) -> Result<
-        (
-            Vec<SpendDescription>,
-            Vec<OutputDescription>,
-            SaplingMetadata,
-        ),
-        Error,
-    > {
+    ) -> Result<(Option<Bundle<Unauthorized>>, SaplingMetadata), Error> {
```
Comment (str4d) [inline zcash_primitives/src/transaction/components/sapling/builder.rs:314]: I think what we actually should return here is:
```suggestion
    ) -> Result<Option<(Bundle<Unauthorized>, SaplingMetadata)>, Error> {
```
since if there is no bundle, there is no metadata.

File: zcash_primitives/src/transaction/components/sapling/builder.rs [hunk 7]
```diff
@@ -491,7 +486,18 @@ impl<P: consensus::Parameters> SaplingBuilder<P> {
...
-        Ok((spend_descs, output_descs, tx_metadata))
+        let bundle = if shielded_spends.is_empty() && shielded_outputs.is_empty() {
+            None
+        } else {
+            Some(Bundle {
+                shielded_spends,
+                shielded_outputs,
+                value_balance: self.value_balance,
+                binding_sig: (),
+            })
```
Comment (str4d) [inline zcash_primitives/src/transaction/components/sapling/builder.rs:497]:
```suggestion
            Some((Bundle {
                shielded_spends,
                shielded_outputs,
                value_balance: self.value_balance,
                binding_sig: (),
            }, tx_metadata))
```

File: zcash_primitives/src/transaction/components/sapling/builder.rs [hunk 8]
```diff
@@ -513,18 +520,100 @@ impl<P: consensus::Parameters> SaplingBuilder<P> {
...
+
+    prop_compose! {
+        fn arb_bundle()(n_notes in 1..30usize)(
+            extsk in arb_extended_spending_key(),
+            spendable_notes in vec(
+                arb_positive_note_value(MAX_MONEY as u64 / 10000).prop_flat_map(arb_note),
+                n_notes
+            ),
+            commitment_trees in vec(
+                arb_commitment_tree(arb_node()).prop_map(
```
Comment (str4d) [inline zcash_primitives/src/transaction/components/sapling/builder.rs:574]: Technically, the tree needs to contain at least as many notes as are being spent, so this might be useful as an `arb_nonempty_commitment_tree(spendable_notes.len())`.

**Review by str4d — COMMENTED** [review]
Partial review.

File: zcash_primitives/src/transaction/sighash.rs [hunk 9]
```diff
@@ -127,7 +104,7 @@ fn joinsplits_hash(
 ) -> Blake2bHash {
     let mut data = Vec::with_capacity(
         joinsplits.len()
-            * if txversion.uses_groth_proofs() {
+            * if txversion.has_sapling() {
```
Comment (str4d) [inline zcash_primitives/src/transaction/sighash.rs:107]: Per my earlier note, this is switching Sprout logic on the presence of Sapling logic, which feels brittle.

File: zcash_primitives/src/transaction/sighash.rs [hunk 10]
```diff
@@ -143,7 +120,9 @@ fn joinsplits_hash(
         .hash(&data)
 }
 
-fn shielded_spends_hash(shielded_spends: &[SpendDescription]) -> Blake2bHash {
+fn shielded_spends_hash<A: sapling::Authorization<Proof = GrothProofBytes>>(
+    shielded_spends: &[SpendDescription<A>],
+) -> Blake2bHash {
```
Comment (str4d) [inline zcash_primitives/src/transaction/sighash.rs:125]: This is so very satisfying.

File: zcash_primitives/src/transaction/mod.rs [hunk 11]
```diff
@@ -90,6 +101,7 @@ pub enum TxVersion {
     Sprout(u32),
     Overwinter,
     Sapling,
+    ZcashTxV5,
```
Comment (str4d) [inline zcash_primitives/src/transaction/mod.rs:103]: This PR adds v5 to `TxVersion`, but it's not intended to actually implement the v5 transaction, correct?

File: zcash_primitives/src/transaction/mod.rs [hunk 12]
```diff
@@ -527,107 +566,49 @@ impl Transaction {
...
+        if let Some(bundle) = &self.tze_bundle {
+            Vector::write(&mut writer, &bundle.vin, |w, e| e.write(w))?;
+            Vector::write(&mut writer, &bundle.vout, |w, e| e.write(w))?;
+        } else {
+            CompactSize::write(&mut writer, 0)?;
+            CompactSize::write(&mut writer, 0)?;
+        }
+
+        Ok(())
+    }
```
Comment (str4d) [inline zcash_primitives/src/transaction/mod.rs:592]: Slightly odd that most TZE transaction format parsing and serialization is removed, but this is left here unused.

**Review by str4d — CHANGES_REQUESTED** [review]

File: zcash_primitives/src/transaction/mod.rs [hunk 13]
```diff
@@ -160,15 +174,17 @@ impl TxVersion {
...
             TxVersion::Sprout(v) => *v >= 2u32,
             TxVersion::Overwinter | TxVersion::Sapling => true,
+            TxVersion::ZcashTxV5 => false,
             #[cfg(feature = "zfuture")]
             TxVersion::ZFuture => true,
         }
     }
 
-    pub fn uses_groth_proofs(&self) -> bool {
+    pub fn has_sapling(&self) -> bool {
```
[the remark this answers is not in the document]
Comment (nuttycom) [reply zcash_primitives/src/transaction/mod.rs:182]: Thinking about this more, the "uses groth proofs" should really be a condition of the consensus branch ID, not of the transaction version.

File: zcash_primitives/src/transaction/sighash.rs (hunk 9 above)
Comment (nuttycom) [reply zcash_primitives/src/transaction/sighash.rs:107]: I'm going to see what it takes to plumb in the consensus branch id here instead.

File: zcash_primitives/src/transaction/components/sapling/builder.rs (hunk 7 above)
Comment (nuttycom) [reply zcash_primitives/src/transaction/components/sapling/builder.rs]: I'm going to defer this until we move signature application into the bundles themselves.

File: zcash_primitives/src/transaction/components/sapling/builder.rs (hunk 6 above)
Comment (nuttycom) [reply zcash_primitives/src/transaction/components/sapling/builder.rs:293]: This is currently an awkward change to make; I'm going to leave it for a future refactoring.

File: zcash_primitives/src/transaction/mod.rs (hunk 11 above)
Comment (nuttycom) [reply zcash_primitives/src/transaction/mod.rs:103]: Correct.

File: zcash_primitives/src/transaction/mod.rs (hunk 12 above)
Comment (nuttycom) [reply zcash_primitives/src/transaction/mod.rs:592]: We no longer have a specification for where the TZE bits of the transaction should go! ZIP-222 needs to be updated, then this code will come back to life.

File: zcash_primitives/src/transaction/builder.rs [hunk 14]
```diff
@@ -358,68 +358,124 @@ impl<'a, P: consensus::Parameters, R: RngCore> Builder<'a, P, R> {
...
+            (Some(bundle), Some(script_sigs)) => Some(bundle.apply_signatures(script_sigs)),
+            (None, None) => None,
+            (b, s) => {
+                panic!(
+                    "Mismatch between transparent bundle ({}) and signatures ({}).",
+                    b.is_some(),
+                    s.is_some()
+                );
+            }
+        };
```
[the remark this answers is not in the document]
Comment (nuttycom) [reply zcash_primitives/src/transaction/builder.rs]: I'm doing this for the next PR (or maybe two down the road, if I split the next one up.)

File: zcash_primitives/src/transaction/components/sapling.rs [hunk 15]
```diff
@@ -1,36 +1,113 @@
...
+                .shielded_spends
+                .iter()
+                .zip(spend_auth_sigs.iter())
+                .map(|(spend, sig)| spend.apply_signature(*sig))
+                .collect(),
+            shielded_outputs: self
+                .shielded_outputs
+                .into_iter()
+                .map(|o| o.into_authorized())
+                .collect(), //TODO, is there a zero-cost way to do this?
```
Comment (str4d) [inline zcash_primitives/src/transaction/components/sapling.rs:80]: Not in the general case. But in this specific case, you could type `OutputDescription` on `T` instead of `A: Authorization` with:
```diff
-    pub zkproof: A::Proof,
+    pub zkproof: T,
```
and then in the definition of `Bundle`:
```diff
-    pub shielded_outputs: Vec<OutputDescription<A>>,
+    pub shielded_outputs: Vec<OutputDescription<A::Proof>>,
```
Then this would become:
```suggestion
            shielded_outputs: self.shielded_outputs,
```

File: zcash_primitives/src/transaction/components/sapling.rs [hunk 16]
```diff
@@ -1,36 +1,113 @@
...
+        spend_auth_sigs: Vec<Signature>,
+        binding_sig: Signature,
+    ) -> Bundle<Authorized> {
+        assert!(self.shielded_spends.len() == spend_auth_sigs.len());
+        Bundle {
+            shielded_spends: self
+                .shielded_spends
+                .iter()
+                .zip(spend_auth_sigs.iter())
+                .map(|(spend, sig)| spend.apply_signature(*sig))
```
Comment (str4d) [inline zcash_primitives/src/transaction/components/sapling.rs:74]: These should all be consuming. It's definitely possible to do within the Rust type system (if necessary by defining a temporary):
```suggestion
                .into_iter()
                .zip(spend_auth_sigs.into_iter())
                .map(|(spend, sig)| spend.apply_signature(sig))
```

File: zcash_primitives/src/transaction/components/sapling.rs [hunk 17]
```diff
@@ -1,36 +1,113 @@
...
 }
 
-impl std::fmt::Debug for SpendDescription {
+impl SpendDescription<Unauthorized> {
+    pub fn apply_signature(&self, spend_auth_sig: Signature) -> SpendDescription<Authorized> {
+        SpendDescription {
+            cv: self.cv,
+            anchor: self.anchor,
+            nullifier: self.nullifier,
+            rk: self.rk.clone(),
```
Comment (str4d) [inline zcash_primitives/src/transaction/components/sapling.rs:103]:
```suggestion
    pub fn apply_signature(self, spend_auth_sig: Signature) -> SpendDescription<Authorized> {
        SpendDescription {
            cv: self.cv,
            anchor: self.anchor,
            nullifier: self.nullifier,
            rk: self.rk,
```

File: zcash_primitives/src/transaction/components/sapling.rs [hunk 18]
```diff
@@ -40,49 +117,84 @@ impl std::fmt::Debug for SpendDescription {
...
+///   (located in zcash_proofs::sapling::verifier).
+pub fn read_point<R: Read>(mut reader: R, field: &str) -> io::Result<jubjub::ExtendedPoint> {
+    let mut bytes = [0u8; 32];
+    reader.read_exact(&mut bytes)?;
+    let point = jubjub::ExtendedPoint::from_bytes(&bytes);
+
+    if point.is_none().into() {
+        Err(io::Error::new(
+            io::ErrorKind::InvalidInput,
+            "invalid ".to_owned() + field,
```
Comment (str4d) [inline zcash_primitives/src/transaction/components/sapling.rs:132]:
```suggestion
            format!("invalid {}", field),
```
This will return a `String`.

File: zcash_primitives/src/transaction/components/sapling.rs [hunk 19]
```diff
@@ -40,49 +117,84 @@ impl std::fmt::Debug for SpendDescription {
...
-        };
+/// Consensus rules (§7.3) & (§7.4):
+/// - Canonical encoding is enforced here
+pub fn read_scalar<R: Read>(mut reader: R, field: &str) -> io::Result<bls12_381::Scalar> {
+    let mut f = [0u8; 32];
+    reader.read_exact(&mut f)?;
+    bls12_381::Scalar::from_repr(f).ok_or_else(|| {
+        io::Error::new(
+            io::ErrorKind::InvalidInput,
+            field.to_owned() + " not in field",
```
Comment (str4d) [inline zcash_primitives/src/transaction/components/sapling.rs:147]:
```suggestion
            format!("{} not in field", field),
```

File: zcash_primitives/src/transaction/components/sapling.rs [hunk 20]
```diff
@@ -40,49 +117,84 @@ impl std::fmt::Debug for SpendDescription {
...
-        // Consensus rule (§7.3): Canonical encoding is enforced here
-        let anchor = {
-            let mut f = [0u8; 32];
-            reader.read_exact(&mut f)?;
-            bls12_381::Scalar::from_repr(f)
-                .ok_or_else(|| io::Error::new(io::ErrorKind::InvalidInput, "anchor not in field"))?
-        };
+/// Consensus rules (§7.3) & (§7.4):
+/// - Canonical encoding is enforced here
+pub fn read_scalar<R: Read>(mut reader: R, field: &str) -> io::Result<bls12_381::Scalar> {
```
Comment (str4d) [inline zcash_primitives/src/transaction/components/sapling.rs:141]: It's confusing to call this `read_scalar`. It's reading a BLS12-381 scalar, which is a Jubjub base field element, and it's the latter interpretation we want (as anchors are extracted from Jubjub curve point coordinates). Meanwhile, we _do_ read Jubjub scalars in other places.

At some point we should add a `type Base = bls12_381::Scalar` to `jubjub`, but in the meantime:
```suggestion
pub fn read_base<R: Read>(mut reader: R, field: &str) -> io::Result<bls12_381::Scalar> {
```

File: zcash_primitives/src/transaction/components/sapling.rs [hunk 21]
```diff
@@ -40,49 +117,84 @@ impl std::fmt::Debug for SpendDescription {
...
+        // Consensus rules (§4.4) & (§4.5):
         // - Canonical encoding is enforced here.
-        // - Signature validity is enforced in SaplingVerificationContext::check_spend()
-        let spend_auth_sig = Some(Signature::read(&mut reader)?);
+        // - "Not small order" is enforced in SaplingVerificationContext::(check_spend()/check_output())
+        //   (located in zcash_proofs::sapling::verifier).
+        let cv = read_point(&mut reader, "cv")?;
+        // Consensus rules (§7.3) & (§7.4):
+        // - Canonical encoding is enforced here
+        let anchor = read_scalar(&mut reader, "anchor")?;
```
Comment (str4d) [inline zcash_primitives/src/transaction/components/sapling.rs:193]:
```suggestion
        let anchor = read_base(&mut reader, "anchor")?;
```

**Review by str4d — APPROVED** [review]
utACK after the `format!` comments are addressed.

File: zcash_primitives/src/transaction/components/sapling.rs (hunk 16 above)
Comment (nuttycom) [reply zcash_primitives/src/transaction/components/sapling.rs:74]: I'm making this irrelevant in https://github.com/zcash/librustzcash/pull/375/commits/a2b6af4230df16838343d34c3a3fdcb6aeb4cef8

File: zcash_primitives/src/transaction/components/sapling.rs (hunk 17 above)
Comment (nuttycom) [reply zcash_primitives/src/transaction/components/sapling.rs:103]: This is also being made irrelevant in https://github.com/zcash/librustzcash/pull/375/commits/a2b6af4230df16838343d34c3a3fdcb6aeb4cef8

File: zcash_primitives/src/transaction/components/sapling.rs (hunk 15 above)
Comment (nuttycom) [reply zcash_primitives/src/transaction/components/sapling.rs:80]: Fixed in https://github.com/zcash/librustzcash/pull/375/commits/1c8fa2ee4e1fa21093b3bc6f5e2052205ff0c170
